### PR TITLE
Find revisions by date and/or username

### PR DESCRIPTION
**Pull request by mathieu-axiocode**
Hi, 
If you think searching revisions by date range or username could be usefull, I have just added a method to your AuditReader to manage this kind of request.

### CODE REVIEW

File: src/SimpleThings/EntityAudit/AuditReader.php [hunk 1]
```diff
@@ -819,4 +819,39 @@ public function getEntityHistory($className, $id)
 
         return $result;
     }
+    
+    public function findRevisionsByDateAndUsername($className, $id, \DateTime $dateFrom = null, \DateTime $dateTo = null, $username = null)
+    {
+        $revisions = array();
+
+        foreach ($this->findRevisions($className, $id) as $revision) {
```
Comment (tolry) [inline src/SimpleThings/EntityAudit/AuditReader.php]: How about using array_filter for this?

File: src/SimpleThings/EntityAudit/AuditReader.php [hunk 2]
```diff
@@ -819,4 +819,39 @@ public function getEntityHistory($className, $id)
 
         return $result;
     }
+    
+    public function findRevisionsByDateAndUsername($className, $id, \DateTime $dateFrom = null, \DateTime $dateTo = null, $username = null)
```
Comment (tolry) [inline src/SimpleThings/EntityAudit/AuditReader.php]: This is quite a large method signature, how about using a criteria object for the additional, optional, filters?

e.g.

```php
<?php

public function matchRevisions($className, $id, RevisionCriteria $criteria)
{
}

// ...

class RevisionCriteria
{
    public $dateFrom;
    public $dateTo;
    public $username;
}
```

File: src/SimpleThings/EntityAudit/AuditReader.php [hunk 3]
```diff
@@ -819,4 +819,39 @@ public function getEntityHistory($className, $id)
...
+        foreach ($this->findRevisions($className, $id) as $revision) {
+            if ($this->revisionMatchDates($revision, $dateFrom, $dateTo) && $this->revisionMatchUsername($revision, $username)) {
+                $revisions[] = $revision;
+            }
+        }
+
+        return $revisions;
+    }
+
+    protected function revisionMatchDates($revision, $dateFrom, $dateTo)
```
Comment (tolry) [inline src/SimpleThings/EntityAudit/AuditReader.php]: missing typehints

File: src/SimpleThings/EntityAudit/AuditReader.php [hunk 4]
```diff
@@ -819,4 +819,39 @@ public function getEntityHistory($className, $id)
...
+        }
+
+        if ((empty($dateFrom) || $revision->getTimeStamp() >= $dateFrom) && (empty($dateTo) || $dateTo >= $revision->getTimeStamp())) {
+            return true;
+        }
+
+        return false;
+    }
+
+    protected function revisionMatchUsername($revision, $username)
```
Comment (tolry) [inline src/SimpleThings/EntityAudit/AuditReader.php]: missing typehint for `$revision`

File: src/SimpleThings/EntityAudit/AuditReader.php (hunk 2 above)
Comment (mathieu-axiocode) [inline src/SimpleThings/EntityAudit/AuditReader.php]: I've added a RevisionCriteria class like you said which contains the array_filter process. Is that what you meant in your comment?

Comment (tolry) [inline src/SimpleThings/EntityAudit/AuditReader.php]: Sorry, my suggestions might have been a bit confusing, sorry for that. Those were two separate comments on seperate parts of the code.

* array_filter was just meant to replace the foreach-loop
* the criteria object should only be a very simple class, a struct/value object with only public properties and no code - this was meant to keep the method signature small + it is easily extendable, just add another property to the criteria and implement the corresponding logic in the repository

File: src/SimpleThings/EntityAudit/AuditReader.php [hunk 5]
```diff
@@ -819,4 +819,39 @@ public function getEntityHistory($className, $id)
...
+    }
+
+    protected function matchCriteria(Revision $revision, RevisionCriteria $criteria)
+    {
+        return ($this->matchDates($revision, $criteria) && $this->matchUsername($revision, $criteria));
+    }
+
+    protected function matchDates(Revision $revision, RevisionCriteria $criteria)
+    {
+        if (empty($criteria->dateFrom) && empty($criteria->dateTo)) {
```
Comment (mathieu-axiocode) [inline src/SimpleThings/EntityAudit/AuditReader.php]: $criteria->dateFrom and ->dateTo  have to be DateTime(), should it be tested?

Comment (tolry) [inline src/SimpleThings/EntityAudit/AuditReader.php]: If you like to have type safety, you will have to use getters/setters for those fields, with the appropriate type hints.

File: src/SimpleThings/EntityAudit/RevisionCriteria.php [hunk 6]
```diff
@@ -0,0 +1,9 @@
+<?php
+namespace SimpleThings\EntityAudit;
+
+class RevisionCriteria
+{
+    protected $dateFrom;
+    protected $dateTo;
+    protected $username;
```
Comment (tolry) [inline src/SimpleThings/EntityAudit/RevisionCriteria.php]: This shouldn't work right now, those properties need to be public in order to access them as above `$criteria->dateFrom`

Comment (mathieu-axiocode) [inline src/SimpleThings/EntityAudit/RevisionCriteria.php]: OK so I left RevisionCriteria properties as protected and add getters and setters to force the type when necessary, is that ok for you?

Comment (tolry) [inline src/SimpleThings/EntityAudit/RevisionCriteria.php]: fine by me :smile:

File: src/SimpleThings/EntityAudit/AuditReader.php [hunk 7]
```diff
@@ -819,4 +819,44 @@ public function getEntityHistory($className, $id)
 
         return $result;
     }
+    
+    public function matchRevisions($className, $id, RevisionCriteria $criteria)
+    {
+        $revisions = $this->findRevisions($className, $id);
+        return array_filter($revisions, function($revision) use ($criteria) {
+            return $this->matchCriteria($revision, $criteria);
```
Comment (smoench) [inline src/SimpleThings/EntityAudit/AuditReader.php]: `$this` isn't working with PHP 5.3 in this context. But PHP 5.3 support should be dropped anyway (IHMO)

**Comment by smoench** [comment]
@mathieu-axiocode Commits should be squashed before merging.
Please add some docblocks and tests for those new methods.

File: src/SimpleThings/EntityAudit/AuditReader.php [hunk 8]
```diff
@@ -819,4 +819,44 @@ public function getEntityHistory($className, $id)
 
         return $result;
     }
+    
+    public function matchRevisions($className, $id, RevisionCriteria $criteria)
```
Comment (smoench) [inline src/SimpleThings/EntityAudit/AuditReader.php]: This method isn't matching any revision but returning matched revision in an new array. `matchRevisions` should be renamed to something more meaningful, such as `findRevisionsByCriteria`

**Pull request closed without merging**